### ISSUE TITLE
S3 SSE-KMS

### ISSUE THREAD

**Issue by iamriel** [jschneier/django-storages#113]
Hi guys, this is my first time opening a ticket in an open source project.  I am trying to implement AWS SSE-KMS in my django project.  I have added these into my settings.py

```
AWS_S3_ENCRYPTION = True
os.environ['S3_USE_SIGV4'] = 'True'
AWS_S3_HOST = 's3.amazonaws.com'
```

I also used this as my media storage:

```
class MediaStorage(S3BotoStorage):
    location = settings.MEDIAFILES_LOCATION
    @property
    def connection(self):
        if self._connection is None:
            self._connection = self.connection_class(
                self.access_key, self.secret_key,
                calling_format=self.calling_format, host=settings.AWS_S3_HOST)
        return self._connection
```

But my problem is it is raising S3ResponseError: 400 Bad Request.  Here is the traceback:

```
Traceback (most recent call last):
  File "/Users/light/.virtualenvs/von/lib/python2.7/site-packages/django/core/handlers/base.py", line 149, in get_response
    response = self.process_exception_by_middleware(e, request)
  File "/Users/light/.virtualenvs/von/lib/python2.7/site-packages/django/core/handlers/base.py", line 147, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/Users/light/.virtualenvs/von/lib/python2.7/site-packages/django/contrib/admin/options.py", line 541, in wrapper
    return self.admin_site.admin_view(view)(*args, **kwargs)
  File "/Users/light/.virtualenvs/von/lib/python2.7/site-packages/django/utils/decorators.py", line 149, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/Users/light/.virtualenvs/von/lib/python2.7/site-packages/django/views/decorators/cache.py", line 57, in _wrapped_view_func
    response = view_func(request, *args, **kwargs)
  File "/Users/light/.virtualenvs/von/lib/python2.7/site-packages/django/contrib/admin/sites.py", line 244, in inner
    return view(request, *args, **kwargs)
  File "/Users/light/.virtualenvs/von/lib/python2.7/site-packages/django/contrib/admin/options.py", line 1438, in change_view
    return self.changeform_view(request, object_id, form_url, extra_context)
  File "/Users/light/.virtualenvs/von/lib/python2.7/site-packages/django/utils/decorators.py", line 67, in _wrapper
    return bound_func(*args, **kwargs)
  File "/Users/light/.virtualenvs/von/lib/python2.7/site-packages/django/utils/decorators.py", line 149, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/Users/light/.virtualenvs/von/lib/python2.7/site-packages/django/utils/decorators.py", line 63, in bound_func
    return func.__get__(self, type(self))(*args2, **kwargs2)
  File "/Users/light/.virtualenvs/von/lib/python2.7/site-packages/django/utils/decorators.py", line 184, in inner
    return func(*args, **kwargs)
  File "/Users/light/.virtualenvs/von/lib/python2.7/site-packages/django/contrib/admin/options.py", line 1378, in changeform_view
    self.save_model(request, new_object, form, not add)
  File "/Users/light/.virtualenvs/von/lib/python2.7/site-packages/django/contrib/admin/options.py", line 991, in save_model
    obj.save()
  File "/Users/light/.virtualenvs/von/lib/python2.7/site-packages/django/db/models/base.py", line 700, in save
    force_update=force_update, update_fields=update_fields)
  File "/Users/light/.virtualenvs/von/lib/python2.7/site-packages/django/db/models/base.py", line 728, in save_base
    updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
  File "/Users/light/.virtualenvs/von/lib/python2.7/site-packages/django/db/models/base.py", line 790, in _save_table
    for f in non_pks]
  File "/Users/light/.virtualenvs/von/lib/python2.7/site-packages/django/db/models/fields/files.py", line 311, in pre_save
    file.save(file.name, file, save=False)
  File "/Users/light/.virtualenvs/von/lib/python2.7/site-packages/django/db/models/fields/files.py", line 93, in save
    self.name = self.storage.save(name, content, max_length=self.field.max_length)
  File "/Users/light/.virtualenvs/von/lib/python2.7/site-packages/django/core/files/storage.py", line 63, in save
    name = self._save(name, content)
  File "/Users/light/.virtualenvs/von/lib/python2.7/site-packages/storages/backends/s3boto.py", line 385, in _save
    key = self.bucket.get_key(encoded_name)
  File "/Users/light/.virtualenvs/von/lib/python2.7/site-packages/boto/s3/bucket.py", line 192, in get_key
    key, resp = self._get_key_internal(key_name, headers, query_args_l)
  File "/Users/light/.virtualenvs/von/lib/python2.7/site-packages/boto/s3/bucket.py", line 230, in _get_key_internal
    response.status, response.reason, '')
S3ResponseError: S3ResponseError: 400 Bad Request
```

Is there anything I missed?  Thanks for your help.

**Comment by laurrentt** [comment]
@iamriel Did you find a solution yet ?

**Comment by iamriel** [comment]
@laurrentt Sorry I haven't found a solution to this.  I resorted to using SSE-S3 for now.

**Comment by laurrentt** [comment]
A few hints: 
 * `AWS_S3_ENCRYPTION = True` overrides the encryption type to `AES256` so you should remove that. 
 * Add this to your settings:
```
AWS_HEADERS = {
    'x-amz-server-side-encryption': 'aws:kms',
    'x-amz-server-side-encryption-aws-kms-key-id': kms key id or arn (I don't know which one for sure)
}
```

Now if you look at this PR: https://github.com/jschneier/django-storages/pull/111, the repo author wrote 11 days ago that this was close to being merged and boto3 has a much better KMS support. I think I'm gonna wait for the PR to get merge.

**Comment by vCra** [comment]
For anyone else stumbling across this, This should get fixed in #457
